packages/table/src/table.tsx

```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         this.scrollBodyToFocusedCell(newFocusedCell);
     }
 
+    // no good way to call arrow-key keyboard events from tests
+    /* istanbul ignore next */
+    private handleFocusMoveInternal = (e: KeyboardEvent, direction: "up" | "down" | "left" | "right") => {
+        e.preventDefault();
+        e.stopPropagation();
+
+        const { focusedCell, selectedRegions } = this.state;
+        const { grid } = this;
+
+        if (focusedCell == null) {
+            // halt early if we have a selectedRegionTransform or something else in play that nixes
+            // the focused cell.
+            return;
+        }
+
+        let newFocusedCell = {
+            col: focusedCell.col,
+            focusSelectionIndex: focusedCell.focusSelectionIndex,
+            row: focusedCell.row,
+        };
+
+        // if we're not in any particular focus cell region, and one exists, go to the first cell of the first one
+        if (focusedCell.focusSelectionIndex == null && selectedRegions.length > 0) {
+            const focusCellRegion = Regions.getCellRegionFromRegion(
+                selectedRegions[0],
+                grid.numRows,
+                grid.numCols,
+            );
+
+            newFocusedCell = {
+                col: focusCellRegion.cols[0],
+                focusSelectionIndex: 0,
+                row:  focusCellRegion.rows[0],
+            };
+        } else {
+            if (selectedRegions.length === 0) {
+                this.handleFocusMove(e, direction);
+                return;
+            }
+
+            const focusCellRegion = Regions.getCellRegionFromRegion(
+                selectedRegions[focusedCell.focusSelectionIndex],
+                grid.numRows,
+                grid.numCols,
+            );
+
+            if (focusCellRegion.cols[0] === focusCellRegion.cols[1] &&
+                focusCellRegion.rows[0] === focusCellRegion.rows[1] &&
+                selectedRegions.length === 1) {
+                this.handleFocusMove(e, direction);
+                return;
+            }
+
+            switch (direction) {
+                case "up":
+                    newFocusedCell.row -= 1;
```

Comment:
For all this repeated code, I goofed around and got to this, which might be more abstracted than it's worth:

```
function moveFocusCell(primaryAxis: "row" | "col", secondaryAxis: "row" | "col", isUpOrLeft: boolean) {
    const primaryAxisPlural = primaryAxis === "row" ? "rows" : "cols";
    const secondaryAxisPlural = secondaryAxis === "row" ? "rows" : "cols";

    const generalDelta = isUpOrLeft ? -1 : +1;
    const generalIndex = isUpOrLeft ? 1 : 0;

    newFocusedCell[primaryAxis] += generalDelta;

    const isPrimaryIndexOutOfBounds = isUpOrLeft
        ? newFocusedCell[primaryAxis] < focusCellRegion[primaryAxisPlural][0]
        : newFocusedCell[primaryAxis] > focusCellRegion[primaryAxisPlural][1];

    if (isPrimaryIndexOutOfBounds) {
        newFocusedCell[primaryAxis] = focusCellRegion[primaryAxisPlural][generalIndex];
        newFocusedCell[secondaryAxis] += generalDelta;

        const isSecondaryIndexOutOfBounds = isUpOrLeft
            ? newFocusedCell[secondaryAxis] < focusCellRegion[secondaryAxisPlural][0]
            : newFocusedCell[secondaryAxis] > focusCellRegion[secondaryAxisPlural][1];

        if (isSecondaryIndexOutOfBounds) {
            let newFocusCellSelectionIndex = focusedCell.focusSelectionIndex + generalDelta;
            
            if (isUpOrLeft
                ? newFocusCellSelectionIndex < 0
                : newFocusCellSelectionIndex >= selectedRegions.length
            ) {
                newFocusCellSelectionIndex = isUpOrLeft
                    ? selectedRegions.length - 1
                    : 0;
            }
            
            const newFocusCellRegion = Regions.getCellRegionFromRegion(
                selectedRegions[newFocusCellSelectionIndex],
                grid.numRows,
                grid.numCols,
            );

            newFocusedCell = {
                col: newFocusCellRegion.cols[generalIndex],
                focusSelectionIndex: newFocusCellSelectionIndex,
                row: newFocusCellRegion.rows[generalIndex],
            };
        }
    }
}
```

You could then call this function as follows:

```
switch (direction) {
    case "up":
        moveFocusCell("row", "col", true);
    case "left":
        moveFocusCell("col", "row", true);
    case "down":
        moveFocusCell("row", "col", false);
    case "right":
        moveFocusCell("col", "row", false);
}
```